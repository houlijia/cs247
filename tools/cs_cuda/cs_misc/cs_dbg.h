#ifndef __CS_DBG_H__ 
#define __CS_DBG_H__ 

#include "cs_header.h"

// #define CUDA_DBG 

int dbg_init( int ) ;
void dbg_clear_buf( int *cp, int size ) ;
void dbg_set_buf( int *cp, int size, int set ) ;
int dbg_put_d_data ( char *dp, char *hp, int size )  ;
int dbg_get_d_data ( char *dp, char *hp, int size )  ;
void dbg_pdata_i( char *s, int *dp, int size ) ;
void dbg_pdata_c( char *s, char *dp, int size ) ;
void dbg_mdata( int *dp, int size ) ;
void dbg_p_d_data_c ( char *s, char *dp, int size ) ;
void dbg_p_d_data_i ( char *s, int *dp, int size ) ;
char *dbg_d_malloc_c ( int size ) ;
int *dbg_d_malloc_i ( int size ) ;
void dbg_p_d_data_i_mn_skip ( char *s, int *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;

void dbg_p_d_data_i_mn ( char *s, int *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_c_mn ( char *s, char *dp, int size, int m, int n, int doprint ) ;
void dbg_p_data_i_mn ( char *s, int *dp, int size, int m, int n, int doprint ) ;

void dbg_p_d_data_i_mn_v2 ( char *s, int *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_data_i_mn_v2 ( char *s, int *hp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;

// tsting ...
void dbg_p_d_data_ll ( char *s, long long *dp, int size ) ;

#endif 
