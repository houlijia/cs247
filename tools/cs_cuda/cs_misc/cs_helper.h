#ifndef __CS_HELPER_H__
#define __CS_HELPER_H__

int max_log2( int ) ;

// clear the device mem ... n in size of int
void clear_device_mem_i ( int *d_a, int n ) ;
void set_device_mem_i ( int *d_a, int n, int val ) ;

// clear the device mem ... n in size of char
void clear_device_mem_c ( char *d_a, int n ) ;
void set_device_mem_c ( char *d_a, int n, char c ) ;
void h_expand_c_to_i ( char *din, int *dout, int n ) ;	

// block adjustment
void h_block_adj( int n, int nThreadsPerBlock, int *block );
//

// omp timer 
int omp_timer_init( int ) ;
int omp_timer_on( int ) ;
int omp_timer_off( int ) ;
int omp_timer_get( int, double *dp, int *cp, double *ddp ) ;

// opt
int get_nums( int ac, char *av[], int idx, int cnt, int *np ) ;
unsigned int htoi(const char s[]) ;

// cs timer 
int cs_timer_init( int cnt ) ;
int cs_timer_on( int idx ) ;
int cs_timer_off( int idx ) ;
int cs_timer_get( int idx, clock_t *sp, clock_t *up, int *cp,
	double *asp, double *aup ) ;

// swap
void htonl_device_mem_i ( int *d_a, int n )	;

// tsting ...
void h_tst_longlong ( long long *d_a, int n )	 ;

// abs a vector

int h_set_abs ( int *d_a, int n, int record_len, int skip )	 ;

// set config
void h_set_config( struct cs_xyz *d_a, struct cube *hp ) ;

// weight shift
int weight_sft( int scheme, int size, int m, int n ) ;

// generic read/write int from/to device ... for ret code
int put_d_data_i ( int *dp, int *hp, int size ) ;
int get_d_data_i ( int *dp, int *hp, int size ) ;

#endif 
