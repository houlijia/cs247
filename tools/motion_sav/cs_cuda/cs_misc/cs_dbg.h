#ifndef __CS_DBG_H__ 
#define __CS_DBG_H__ 

#include "cs_header.h"

// #define CUDA_DBG 

int dbg_init( int ) ;
void dbg_clear_buf( int *cp, int size ) ;
void dbg_set_buf( int *cp, int size, int set ) ;
int dbg_put_d_data ( char *dp, char *hp, int size )  ;
int dbg_get_d_data ( char *dp, char *hp, int size )  ;
int dbg_copy_d_data ( char *dtp, char *dfp, int size ) ;
void dbg_pdata_i( const char *s, const int *dp, int size ) ;
void dbg_pdata_f( const char *s, const float *dp, int size ) ;
void dbg_pdata_c( const char *s, const char *dp, int size ) ;
void dbg_pdata_d( const char *s, const float *dp, int size ) ;
void dbg_mdata( int *dp, int size ) ;
void dbg_p_d_data_c ( const char *s, char *dp, int size ) ;
void dbg_p_d_data_i ( const char *s, int *dp, int size ) ;
void dbg_p_d_data_f ( const char *s, float *dp, int size ) ;
void dbg_p_d_data_d ( const char *s, float *dp, int size ) ;
char *dbg_d_malloc_c ( int size ) ;
int *dbg_d_malloc_i ( int size ) ;
float *dbg_d_malloc_f ( int size ) ;

void dbg_p_d_data_f_mn_skip ( const char *s, float *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;
void dbg_p_d_data_i_mn_skip ( const char *s, int *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;

void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint, int printrow ) ;
void dbg_p_d_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_c_mn ( const char *s, char *dp, int size, int m, int n, int doprint ) ;
void dbg_p_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

// print the motion detection T/V/H/Va format
void dbg_p_data_md_f_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

void dbg_p_d_data_f_mn_v2 ( const char *s, float *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_d_data_i_mn_v2 ( const char *s, int *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_data_i_mn_v2 ( const char *s, int *hp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;

void dbg_p_d_data_i_cube ( const char *s, int *devp, int vx, int hy, int tz ) ;
void dbg_p_d_data_i_cube ( const char *s, float *devp, int vx, int hy, int tz ) ;
void dbg_p_data_i_cube ( const char *s, float *dp, int vx, int hy, int tz ) ;

void dbg_pr_first_last ( char *s, float *d_p, int len, int pr_size ) ;
void dbg_pr_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, char *h_p, int len, int pr_size ) ;

// tsting ...
void dbg_p_d_data_ll ( const char *s, long long *dp, int size ) ;

// for random
int dbg_ck_unique ( char *s, int *dp, int size ) ;
int dbg_perm_ck ( int callid, int callid2, int zero_in_zero, int *d_bp, int size ) ;

// TVH
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print, int do_skip ) ;
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print,
	int do_skip, int start ) ;

#endif 
