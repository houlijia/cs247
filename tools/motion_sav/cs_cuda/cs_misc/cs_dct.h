#ifndef __CS_DCT_H__
#define __CS_DCT_H__

int h_do_dct ( int *d_input, int *d_output, int xdim, int ydim, int zdim ) ;
int h_do_dct ( float *d_input, float *d_output, int xdim, int ydim, int zdim, int inverse ) ;

int h_do_dct_init () ;

#endif 
