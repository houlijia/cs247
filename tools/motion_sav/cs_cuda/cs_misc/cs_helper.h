#ifndef __CS_HELPER_H__
#define __CS_HELPER_H__

int max_log2( int ) ;

// clear the device mem ... n in size of int
void set_device_mem_i ( int *d_a, int n, int val ) ;
void clear_device_mem_i ( int *d_a, int n ) ;

void set_device_mem ( float *d_a, int n, float val ) ;
void clear_device_mem ( float *d_a, int n )	 ;

// clear the device mem ... n in size of char
void clear_device_mem_c ( char *d_a, int n ) ;
void set_device_mem_c ( char *d_a, int n, char c ) ;

void h_expand_c_to_i ( char *din, int *dout, int n ) ;	
void h_expand_c_to_i ( unsigned char *din, int *dout, int n ) ;	
template<typename F, typename T> void h_cast_T_to_T ( F *din, T *dout, int n ) ;

// block adjustment
void h_block_adj( int n, int nThreadsPerBlock, int *block );
//

// omp timer 
int omp_timer_init( int ) ;
int omp_timer_on( int ) ;
int omp_timer_off( int ) ;
int omp_timer_get( int, double *dp, int *cp, double *ddp ) ;

// opt
int get_nums( int ac, char *av[], int idx, int cnt, int *np ) ;
unsigned int htoi(const char s[]) ;

// cs timer 
int cs_timer_init( int cnt ) ;
int cs_timer_on( int idx ) ;
int cs_timer_off( int idx ) ;
int cs_timer_get( int idx, clock_t *sp, clock_t *up, int *cp,
	double *asp, double *aup ) ;

// swap
void htonl_device_mem_i ( int *d_a, int n )	;

// tsting ...
void h_tst_longlong ( long long *d_a, int n )	 ;

// abs a vector

int h_set_abs ( int *d_a, int n, int record_len, int skip )	 ;

// set config
void h_set_config( struct cs_xyz *d_a, struct cube *hp ) ;
int h_set_cube_config( struct cube *d_a, struct cube *hp ) ;
void h_p_d_cube_config ( const char *, struct cube *dp ) ;
void h_p_h_cube_config ( const char *, struct cube *dp ) ;

// weight shift
int weight_sft( int scheme, int size, int m, int n ) ;

// generic read/write int from/to device ... for ret code
int put_d_data_i ( int *dp, int *hp, int size ) ;
int put_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes
int get_d_data_i ( int *dp, int *hp, int size ) ;
int get_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes

void h_do_int_to_float ( int *intp, float *fp, int tbl_size ) ;

#endif 
