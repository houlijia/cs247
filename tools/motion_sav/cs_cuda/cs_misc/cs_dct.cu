#include <stdio.h>
#include <stdlib.h>
#include "cs_cuda.h"
#include "cs_helper.h"
#include "cs_dbg.h"
#include "cs_dct.h"

#define  CUDA_DBG

// using the DCT to perform the sparsifying transformation

static int init_done = 0 ;

#define MIN_DCT_TBL_SIZE	2
#define MAX_DCT_TBL_SIZE	5

float dct2[]= { 
	0.7071, 0.7071, 
	0.7071, -0.7071 } ;

float dct3[]= {
	0.5774, 0.5774, 0.5774,
	0.7071, 0.0000, -0.7071,
	0.4082, -0.8165, 0.4082 } ;

float dct4[]= {
	0.5000,    0.5000,    0.5000,    0.5000,
	0.6533,    0.2706,   -0.2706,   -0.6533,
	0.5000,   -0.5000,   -0.5000,    0.5000,
	0.2706,   -0.6533,    0.6533,   -0.2706 } ;

float dct5[]= {
	0.4472,    0.4472,    0.4472,    0.4472,    0.4472,
	0.6015,    0.3717,    0.0000,   -0.3717,   -0.6015,
	0.5117,   -0.1954,   -0.6325,   -0.1954,    0.5117,
	0.3717,   -0.6015,   -0.0000,    0.6015,   -0.3717,
	0.1954,   -0.5117,    0.6325,   -0.5117,    0.1954 } ;

static float *dct_result ; // double as idct_result
static float *dct2_dp ;
static float *dct3_dp ;
static float *dct4_dp ;
static float *dct5_dp ;

float *dcttab[4] ;

float idct2[]= { 
	0.7071, 0.7071, 
	0.7071, -0.7071 } ;

float idct3[]= {
	0.5774,    0.7071,    0.4082,
	0.5774,         0,   -0.8165,
	0.5774,   -0.7071,    0.4082 } ;

float idct4[]= {
	0.5000,    0.6533,    0.5000,    0.2706,
	0.5000,    0.2706,   -0.5000,   -0.6533,
	0.5000,   -0.2706,   -0.5000,    0.6533,
	0.5000,   -0.6533,    0.5000,   -0.2706 } ;

float idct5[]= {
	0.4472,    0.6015,    0.5117,    0.3717,    0.1954,
	0.4472,    0.3717,   -0.1954,   -0.6015,   -0.5117,
	0.4472,         0,   -0.6325,    0.0000,    0.6325,
	0.4472,   -0.3717,   -0.1954,    0.6015,   -0.5117,
	0.4472,   -0.6015,    0.5117,   -0.3717,    0.1954 } ;

// static float *idct_result ;
static float *idct2_dp ;
static float *idct3_dp ;
static float *idct4_dp ;
static float *idct5_dp ;

float *idcttab[4] ;

/*
d_do_dcta :::

fdp : from data pointer ... in d mem
tdp : to data pointer ... in d mem
tbl_size : fdp/tdp data size ( in number of int )
dctdp : dct table from one of the above
cx, cy, cz : the data format ... of *fdp, *tdp
	cz is also the size, one dimension, of the dct table
*/

__global__ void d_do_dct ( int *fdp, int *tdp, int tbl_size, float *dctdp,
	int cx, int cy, int cz, float *dct_d_result )
{
	int i, j, k, *fp, *tp ;
	float *dp, d ;
	int xy = cx * cy ;

	int ot_idx, t_idx = blockIdx.x*blockDim.x + threadIdx.x;

	// the size is the total size on device

	while ( t_idx < tbl_size )
	{
		ot_idx = t_idx ;

		/// old starts from here ... ldl ... 06/05/14

		i = t_idx / xy ;
		j = t_idx % xy ;

		dp = dctdp + i * cz ;
	   	// fp = fdp + i * xy + j ;
	   	fp = fdp + j ;
	   	tp = tdp + i * xy + j ;

		d = 0.0 ;
		for ( k = 0 ; k < cz ; k++ )
		{
			d += ( *dp ) * ((float)*fp) ;
			dp++ ;
			fp += xy ;
		}
			
		*tp = roundf(d) ;

		t_idx = ot_idx + CUDA_MAX_THREADS ;
	} 
#ifdef CUDA_OBS 
	else
		cudadbgp[ 64 ] = t_idx ;
#endif 

	*dct_d_result = 1.0 ; // good
}

__global__ void d_do_dct ( float *fdp, float *tdp, int tbl_size, float *dctdp,
	int cx, int cy, int cz, float *dct_d_result )
{
	int i, j, k ;
	float *fp, *tp, *dp, d ;
	int xy = cx * cy ;

	int ot_idx, t_idx = blockIdx.x*blockDim.x + threadIdx.x;

	// the size is the total size on device

	while ( t_idx < tbl_size )
	{
		ot_idx = t_idx ;

		/// old starts from here ... ldl ... 06/05/14

		i = t_idx / xy ;
		j = t_idx % xy ;

		dp = dctdp + i * cz ;
	   	// fp = fdp + i * xy + j ;
	   	fp = fdp + j ;
	   	tp = tdp + i * xy + j ;

		d = 0.0 ;
		for ( k = 0 ; k < cz ; k++ )
		{
			d += ( *dp ) * ((float)*fp) ;
			dp++ ;
			fp += xy ;
		}
			
		*tp = d ;

		t_idx = ot_idx + CUDA_MAX_THREADS ;
	} 
#ifdef CUDA_OBS 
	else
		cudadbgp[ 64 ] = t_idx ;
#endif 

	*dct_d_result = 1.0 ; // good
}

/* 
input : device addr
output : device addr
xdim : x dimension of frame 
ydim : y dimension of frame
zdim : z dimension of frame // time 

return: 0: fail 1: good
*/
int
h_do_dct ( int *d_input, int *d_output, int xdim, int ydim, int zdim, int inverse )
{
	int nThreadsPerBlock = CUDA_MAX_THREADS_P_BLK ;
	float *dct_tbl_p ;
	int nn = xdim * ydim * zdim ;
	int nBlocks ; // = ( nn + ( nThreadsPerBlock - 1 ))/nThreadsPerBlock ;

	if ( !init_done )
	{
		fprintf( stderr, "%s: init not done\n", __func__ ) ;
		return ( 0 ) ;
	}


	if (( zdim > MAX_DCT_TBL_SIZE ) || ( zdim < MIN_DCT_TBL_SIZE )) 
		return ( 0 ) ;

	if ( inverse )
		dct_tbl_p = idcttab[ zdim - MIN_DCT_TBL_SIZE ] ;
	else
		dct_tbl_p = dcttab[ zdim - MIN_DCT_TBL_SIZE ] ;

#ifdef CUDA_DBG 
	fprintf( stderr, "%s: din %p dout %p x/y/z %d %d %d dct_p %p nn %d\n",
		__func__,
		d_input, d_output, xdim, ydim, zdim, dct_tbl_p, nn ) ;
#endif 

	if (( zdim > MAX_DCT_TBL_SIZE ) || ( zdim < MIN_DCT_TBL_SIZE )) 
		return ( 0 ) ;

	// dbg_put_d_data ( d_dct_tbl_p, dct_tbl_p[ zdim - MIN_DCT_TBL_SIZE ] );

	h_block_adj ( nn, nThreadsPerBlock, &nBlocks ) ;

	d_do_dct <<< nBlocks, nThreadsPerBlock >>> ( d_input, d_output,
		xdim * ydim * zdim, dct_tbl_p, xdim, ydim, zdim, dct_result ) ;

	cudaThreadSynchronize() ;

	return ( 1 ) ;
}

int
h_do_dct ( float *d_input, float *d_output, int xdim, int ydim, int zdim, int inverse )
{
	int nThreadsPerBlock = CUDA_MAX_THREADS_P_BLK ;
	float *dct_tbl_p ;
	int nn = xdim * ydim * zdim ;
	int nBlocks ; // = ( nn + ( nThreadsPerBlock - 1 ))/nThreadsPerBlock ;

	if ( !init_done )
	{
		fprintf( stderr, "%s: init not done\n", __func__ ) ;
		return ( 0 ) ;
	}

	if (( zdim > MAX_DCT_TBL_SIZE ) || ( zdim < MIN_DCT_TBL_SIZE )) 
		return ( 0 ) ;

	if ( inverse )
		dct_tbl_p = idcttab[ zdim - MIN_DCT_TBL_SIZE ] ;
	else
		dct_tbl_p = dcttab[ zdim - MIN_DCT_TBL_SIZE ] ;

#ifdef CUDA_DBG 
	fprintf( stderr, "%s: din %p dout %p x/y/z %d %d %d dct_p %p nn %d\n",
		__func__,
		d_input, d_output, xdim, ydim, zdim, dct_tbl_p, nn ) ;
#endif 

	if (( zdim > MAX_DCT_TBL_SIZE ) || ( zdim < MIN_DCT_TBL_SIZE )) 
		return ( 0 ) ;

	// dbg_put_d_data ( d_dct_tbl_p, dct_tbl_p[ zdim - MIN_DCT_TBL_SIZE ] );

	h_block_adj ( nn, nThreadsPerBlock, &nBlocks ) ;

	d_do_dct <<< nBlocks, nThreadsPerBlock >>> ( d_input, d_output,
		xdim * ydim * zdim, dct_tbl_p, xdim, ydim, zdim, dct_result ) ;

	cudaThreadSynchronize() ;

	return ( 1 ) ;
}

int
h_do_dct_init ()
{
	float *dp ;
	int j, i ;

	i = sizeof ( dct2 ) +
		sizeof ( dct3 ) +
		sizeof ( dct4 ) +
		sizeof ( dct5 )  ;

	if (( j = cudaMalloc( &dp, i * 2 + 100 )) != cudaSuccess )
	{
		fprintf( stderr, "%s: cudaMalloc failed %d\n", __func__, j ) ;
		return ( 0 ) ;
	}

	printf("%s: dct table %lx, size %d -- %lu \n", __func__,
	       (unsigned long)dp, i, (unsigned long)(i / sizeof ( float ))) ;

	dct_result = dp++ ;

	dct2_dp = dp ;
	cudaMemcpy( dp, dct2, sizeof ( dct2 ), cudaMemcpyHostToDevice ) ;
	dp += sizeof ( dct2 ) / sizeof ( float ) ;

	dct3_dp = dp ;
	cudaMemcpy( dp, dct3, sizeof ( dct3 ), cudaMemcpyHostToDevice ) ;
	dp += sizeof ( dct3 ) / sizeof ( float ) ;

	dct4_dp = dp ;
	cudaMemcpy( dp, dct4, sizeof ( dct4 ), cudaMemcpyHostToDevice ) ;
	dp += sizeof ( dct4 ) / sizeof ( float ) ;

	dct5_dp = dp ;
	cudaMemcpy( dp, dct5, sizeof ( dct5 ), cudaMemcpyHostToDevice ) ;

	dcttab[0] = dct2_dp ;
	dcttab[1] = dct3_dp ;
	dcttab[2] = dct4_dp ;
	dcttab[3] = dct5_dp ;

#ifdef CUDA_OBS 
	dbg_p_d_data_d ("h_do_dct_init:2", dcttab[0], sizeof ( dct2 )) ;
	dbg_p_d_data_d ("h_do_dct_init:3", dcttab[1], sizeof ( dct3 )) ;
	dbg_p_d_data_d ("h_do_dct_init:4", dcttab[2], sizeof ( dct4 )) ;
	dbg_p_d_data_d ("h_do_dct_init:5", dcttab[3], sizeof ( dct5 )) ;
#endif 

	// idct tables 

	dp += sizeof ( dct5 ) / sizeof ( float ) ;

	// idct_result = dp++ ;

	idct2_dp = dp ;
	cudaMemcpy( dp, idct2, sizeof ( idct2 ), cudaMemcpyHostToDevice ) ;
	dp += sizeof ( idct2 ) / sizeof ( float ) ;

	idct3_dp = dp ;
	cudaMemcpy( dp, idct3, sizeof ( idct3 ), cudaMemcpyHostToDevice ) ;
	dp += sizeof ( idct3 ) / sizeof ( float ) ;

	idct4_dp = dp ;
	cudaMemcpy( dp, idct4, sizeof ( idct4 ), cudaMemcpyHostToDevice ) ;
	dp += sizeof ( idct4 ) / sizeof ( float ) ;

	idct5_dp = dp ;
	cudaMemcpy( dp, idct5, sizeof ( idct5 ), cudaMemcpyHostToDevice ) ;

	idcttab[0] = idct2_dp ;
	idcttab[1] = idct3_dp ;
	idcttab[2] = idct4_dp ;
	idcttab[3] = idct5_dp ;

#ifdef CUDA_OBS 
	dbg_p_d_data_d ("h_do_dct_init:2", idcttab[0], sizeof ( idct2 )) ;
	dbg_p_d_data_d ("h_do_dct_init:3", idcttab[1], sizeof ( idct3 )) ;
	dbg_p_d_data_d ("h_do_dct_init:4", idcttab[2], sizeof ( idct4 )) ;
	dbg_p_d_data_d ("h_do_dct_init:5", idcttab[3], sizeof ( idct5 )) ;
#endif 

	init_done = 1 ;
	
	return ( 1 ) ;
}	
