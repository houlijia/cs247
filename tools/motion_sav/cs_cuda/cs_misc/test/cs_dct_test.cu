#include <iostream>
using namespace std;

#include <thrust/reduce.h>
#include <thrust/sequence.h>
#include <thrust/host_vector.h>
#include <thrust/device_vector.h>

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <math.h>

#include "cs_dbg.h"
#include "cs_helper.h"
#include "cs_dct.h"

#define CUDA_DBG

float *dp1 = NULL, *dp2 = NULL ;

#define NUM_OF_HVT_INDEX 3

#define BUF_SIZE	( 1024 * 1024 )
#define BUF_SIZE_INT	( BUF_SIZE * sizeof (int))

#define X_SIZE		3
#define Y_SIZE		4	
#define T_SIZE		3
#define MTRX_SIZE	( X_SIZE * Y_SIZE * T_SIZE )

float randi_100_2_6[] = {		// 2x3x2
	82.0, 13.0, 64.0, 
	28.0, 96.0, 16.0, 
	91.0, 92.0, 10.0,
	55.0, 97.0, 98.0 } ;

/* dct of randi_100_2_6 

122.3295   74.2462   52.3259   
58.6899  136.4716   80.6102

-6.3640  -55.8614   38.1838  
-19.0919   -0.7071  -57.9828

*/

float randi_100_3_12[] {	// 3 x 4 x 3
	69.0,     9.0,    16.0,   100.0,    11.0,    78.0,     9.0,    81.0,    19.0,    14.0,    55.0,    63.0,  
	75.0,    23.0,    83.0,     8.0,    97.0,    82.0,    40.0,    44.0,    27.0,    87.0,    15.0,    36.0,  
	46.0,    92.0,    54.0,    45.0,     1.0,    87.0,    26.0,    92.0,    15.0,    58.0,    86.0,    52.0 
} ;	


/*
   dct of randi_100_3_12[] 
109.6966   71.5914   88.3346   88.3346   62.9312  142.6055   43.3013  125.2850   35.2184   91.7987
16.2635  -58.6899  -26.8701   38.8909    7.0711   -6.3640  -12.0208   -7.7782    2.8284  -31.1127
-14.2887   22.4537  -39.1918   52.6640  -74.3012    0.4082  -18.3712   34.7011   -8.1650  -41.6413

90.0666   87.1799
-21.9203    7.7782
45.3156   17.5547
*/



float buf1[MTRX_SIZE] ; 

struct cs_xyz	hcube[ CUBE_INFO_CNT ], *dcubep ;
struct cube cubecube[ CUBE_INFO_CNT ] ;

int
main( int ac, char *av[] )
{
	int k ;
	float *fp, *dp ;

	setbuf( stdout, NULL ) ;
	setbuf( stderr, NULL ) ;

	dbg_init ( 1024 * 1024 ) ;
	h_do_dct_init() ;

	if (( k = cudaMalloc( &dp1, BUF_SIZE_INT )) != cudaSuccess )
	{
		printf("%s: cube alloc failed %d \n", __func__, k ) ;
		exit ( 0 ) ;
	}

	if (( k = cudaMalloc( &dp2, BUF_SIZE_INT )) != cudaSuccess )
	{
		printf("%s: cube alloc failed %d \n", __func__, k ) ;
		exit ( 0 ) ;
	}

	if ( !h_do_dct_init())
	{
		printf("dct init failed") ;
		exit ( 0 ) ;
	}

	dp = buf1 ;

#ifdef CUDA_DBG 
	// setup data for 2x3x2 case ...

	// fp = randi_100_2_6 ;
	fp = randi_100_3_12 ;
	k = MTRX_SIZE ;
	for ( k = 0 ; k < MTRX_SIZE ; k++ )
		*dp++ = *fp++ ;
#endif 

	dbg_put_d_data (( char *)dp1, ( char *)buf1, sizeof( float ) * MTRX_SIZE ) ; 

	// dct

	dbg_p_d_data_i_cube ( "before dct", dp1, X_SIZE, Y_SIZE, T_SIZE ) ;

	if ( !h_do_dct( dp1, dp2, X_SIZE, Y_SIZE, T_SIZE, 0 ))
	{
		printf("%s: h_do_dct failed \n", av[0] ) ;
		exit ( 3 ) ;
	}
  
	dbg_p_d_data_i_cube ( "after dct", dp2, X_SIZE, Y_SIZE, T_SIZE ) ;

	// idct 

	dbg_p_d_data_i_cube ( "before idct", dp1, X_SIZE, Y_SIZE, T_SIZE ) ;

	if ( !h_do_dct( dp1, dp2, X_SIZE, Y_SIZE, T_SIZE, 1 ))	// inverse
	{
		printf("%s: h_do_dct failed \n", av[0] ) ;
		exit ( 3 ) ;
	}
  
	dbg_p_d_data_i_cube ( "after idct", dp2, X_SIZE, Y_SIZE, T_SIZE ) ;

}
