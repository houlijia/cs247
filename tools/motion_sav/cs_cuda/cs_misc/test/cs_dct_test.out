h_do_dct_init: dct table 200000, size 216 -- 54 
h_do_dct_init: dct table 200400, size 216 -- 54 
dbg_p_data_i_cube float: before dct x/y/t 3 4 3
T = 0 
T 0, X 0 
 69.0000   9.0000  16.0000 100.0000 

T 0, X 1 
 11.0000  78.0000   9.0000  81.0000 

T 0, X 2 
 19.0000  14.0000  55.0000  63.0000 


T = 1 
T 1, X 0 
 75.0000  23.0000  83.0000   8.0000 

T 1, X 1 
 97.0000  82.0000  40.0000  44.0000 

T 1, X 2 
 27.0000  87.0000  15.0000  36.0000 


T = 2 
T 2, X 0 
 46.0000  92.0000  54.0000  45.0000 

T 2, X 1 
  1.0000  87.0000  26.0000  92.0000 

T 2, X 2 
 15.0000  58.0000  86.0000  52.0000 


h_do_dct: din 0x1300300000 dout 0x1300700000 x/y/z 3 4 3 dct_p 0x1300200414 nn 36
dbg_p_data_i_cube float: after dct x/y/t 3 4 3
T = 0 
T 0, X 0 
109.7060  71.5976  88.3422  88.3422 

T 0, X 1 
 62.9366 142.6178  43.3050 125.2958 

T 0, X 2 
 35.2214  91.8066  90.0744  87.1874 


T = 1 
T 1, X 0 
 16.2633 -58.6893 -26.8698  38.8905 

T 1, X 1 
  7.0710  -6.3639 -12.0207  -7.7781 

T 1, X 2 
  2.8284 -31.1124 -21.9201   7.7781 


T = 2 
T 2, X 0 
-14.2945  22.4487 -39.1955  52.6570 

T 2, X 1 
-74.3021   0.4000 -18.3730  34.6926 

T 2, X 2 
 -8.1667 -41.6451  45.3087  17.5490 


dbg_p_data_i_cube float: before idct x/y/t 3 4 3
T = 0 
T 0, X 0 
 69.0000   9.0000  16.0000 100.0000 

T 0, X 1 
 11.0000  78.0000   9.0000  81.0000 

T 0, X 2 
 19.0000  14.0000  55.0000  63.0000 


T = 1 
T 1, X 0 
 75.0000  23.0000  83.0000   8.0000 

T 1, X 1 
 97.0000  82.0000  40.0000  44.0000 

T 1, X 2 
 27.0000  87.0000  15.0000  36.0000 


T = 2 
T 2, X 0 
 46.0000  92.0000  54.0000  45.0000 

T 2, X 1 
  1.0000  87.0000  26.0000  92.0000 

T 2, X 2 
 15.0000  58.0000  86.0000  52.0000 


h_do_dct: din 0x1300300000 dout 0x1300700000 x/y/z 3 4 3 dct_p 0x13002004ec nn 36
dbg_p_data_i_cube float: after idct x/y/t 3 4 3
T = 0 
T 0, X 0 
111.6503  59.0143  89.9705  81.7658 

T 0, X 1 
 75.3483 138.5328  44.0938 115.4362 

T 0, X 2 
 36.1853  93.2769  77.4687  83.0582 


T = 1 
T 1, X 0 
  2.2816 -69.9214 -34.8526  20.9975 

T 1, X 1 
  5.5349 -25.9983 -16.0324 -28.3486 

T 1, X 2 
 -1.2769 -39.2734 -38.4620  -6.0818 


T = 2 
T 2, X 0 
  5.5853  26.4877 -27.4081  70.4522 

T 2, X 1 
-61.8291  22.5684 -12.4742  53.2114 

T 2, X 2 
 -1.9981 -29.7585  56.2557  32.1470 


